File: python/examples/transformer_web_interface.py
#!/usr/bin/env python3
"""
LOOM Python Transformer Web Interface
Similar to cabi/web_interface.py but uses the welvet pip package
"""

from http.server import HTTPServer, SimpleHTTPRequestHandler
from urllib.parse import urlparse
import json
import os
import sys
import welvet

# Global state
model_loaded = False
model_path = None


class CORSHTTPRequestHandler(SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        super().end_headers()
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.end_headers()
    
    def do_GET(self):
        parsed = urlparse(self.path)
        
        if parsed.path == '/health':
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            response = {
                'status': 'ok',
                'model': os.path.basename(model_path) if model_path else 'none',
                'model_loaded': model_loaded,
                'backend': 'welvet-python'
            }
            self.wfile.write(json.dumps(response).encode())
        elif parsed.path == '/' or parsed.path == '/inference.html':
            # Serve the HTML file
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            with open('inference.html', 'r') as f:
                self.wfile.write(f.read().encode())
        else:
            super().do_GET()
    
    def do_POST(self):
        if self.path == '/generate':
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            
            try:
                data = json.loads(post_data)
                prompt = data.get('prompt', '')
                max_tokens = data.get('max_tokens', 50)
                temperature = data.get('temperature', 0.7)
                
                if not model_loaded:
                    raise Exception("Model not loaded")
                
                # Set up SSE (Server-Sent Events) for streaming
                self.send_response(200)
                self.send_header('Content-type', 'text/event-stream')
                self.send_header('Cache-Control', 'no-cache')
                self.send_header('Connection', 'keep-alive')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                
                # Stream tokens
                token_count = 0
                for token in welvet.generate_stream(prompt, max_tokens, temperature):
                    event_data = {
                        'token': token,
                        'index': token_count,
                    }
                    self.wfile.write(f"data: {json.dumps(event_data)}\n\n".encode('utf-8'))
                    self.wfile.flush()
                    token_count += 1
                
                # Send completion event
                self.wfile.write(f"data: {json.dumps({'done': True})}\n\n".encode('utf-8'))
                self.wfile.flush()
                
            except Exception as e:
                error_msg = f"data: {json.dumps({'error': str(e)})}\n\n"
                self.wfile.write(error_msg.encode('utf-8'))


def load_model(model_dir):
    """Load transformer model and tokenizer."""
    global model_loaded, model_path
    
    model_path = model_dir
    config_path = os.path.join(model_dir, 'config.json')
    tokenizer_path = os.path.join(model_dir, 'tokenizer.json')
    weights_path = os.path.join(model_dir, 'model.safetensors')
    
    print(f"Loading model from: {model_dir}")
    
    # Load tokenizer
    print("Loading tokenizer...")
    with open(tokenizer_path, 'rb') as f:
        tok_data = f.read()
    
    result = welvet.load_tokenizer_from_bytes(tok_data)
    print(f"✓ Tokenizer loaded (vocab: {result['vocab_size']})")
    
    # Load transformer
    print("Loading transformer model...")
    with open(config_path, 'rb') as f:
        config_data = f.read()
    with open(weights_path, 'rb') as f:
        weights_data = f.read()
    
    print(f"  Config: {len(config_data)} bytes")
    print(f"  Weights: {len(weights_data) / (1024*1024):.2f} MB")
    
    result = welvet.load_transformer_from_bytes(config_data, weights_data)
    print(f"✓ Transformer loaded")
    print(f"  Vocab: {result['vocab_size']}")
    print(f"  Hidden: {result['hidden_size']}")
    print(f"  Layers: {result['num_layers']}")
    
    model_loaded = True


def main():
    if len(sys.argv) < 2:
        print("Usage: ./transformer_web_interface.py <model_path> [port]")
        print("Example: ./transformer_web_interface.py ../../models/SmolLM2-135M-Instruct 8080")
        sys.exit(1)
    
    model_dir = sys.argv[1]
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 8080
    
    # Load model
    try:
        load_model(model_dir)
    except Exception as e:
        print(f"❌ Failed to load model: {e}")
        sys.exit(1)
    
    # Start server
    server_address = ('', port)
    httpd = HTTPServer(server_address, CORSHTTPRequestHandler)
    
    print(f"\n🚀 LOOM Python Transformer Web Interface")
    print(f"   Model: {os.path.basename(model_dir)}")
    print(f"   Server: http://localhost:{port}")
    print(f"   Backend: welvet (pip package)")
    print()
    print("Press Ctrl+C to stop")
    print()
    
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n\n👋 Shutting down...")
        httpd.shutdown()


if __name__ == '__main__':
    main()

File: wrapper/csharp/examples/TransformerWebInterface.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Welvet;

namespace Welvet.Examples;

/// <summary>
/// HTTP server with streaming transformer inference
/// </summary>
public class TransformerWebInterface
{
    private static string modelDirectory = "";

    public static async Task Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Usage: dotnet run <model_directory> <port>");
            Console.WriteLine("Example: dotnet run ../../models/SmolLM2-135M-Instruct 8080");
            return;
        }

        modelDirectory = args[0];
        int port = int.Parse(args[1]);

        Console.WriteLine("Loading model...");

        // Load tokenizer
        string tokenizerPath = Path.Combine(modelDirectory, "tokenizer.json");
        var tokResult = Transformer.LoadTokenizer(tokenizerPath);
        if (!tokResult.Success)
        {
            Console.WriteLine($"Failed to load tokenizer: {tokResult.Error}");
            return;
        }
        Console.WriteLine($"✓ Tokenizer loaded\n  Vocab: {tokResult.VocabSize}");

        // Load model
        var modelResult = Transformer.LoadModelFromDirectory(modelDirectory);
        if (!modelResult.Success)
        {
            Console.WriteLine($"Failed to load model: {modelResult.Error}");
            return;
        }
        Console.WriteLine($"✓ Transformer loaded\n  Vocab: {modelResult.VocabSize}\n  Hidden: {modelResult.HiddenSize}\n  Layers: {modelResult.NumLayers}");

        // Start HTTP server
        Console.WriteLine($"\n🚀 LOOM C# Transformer Web Interface");
        Console.WriteLine($"   Model: {Path.GetFileName(modelDirectory)}");
        Console.WriteLine($"   Server: http://localhost:{port}");
        Console.WriteLine($"   Backend: Welvet (NuGet package)\n");
        Console.WriteLine("Press Ctrl+C to stop\n");

        HttpListener listener = new HttpListener();
        listener.Prefixes.Add($"http://localhost:{port}/");
        listener.Start();

        while (true)
        {
            try
            {
                HttpListenerContext context = await listener.GetContextAsync();
                _ = Task.Run(() => HandleRequest(context));
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
            }
        }
    }

    private static async Task HandleRequest(HttpListenerContext context)
    {
        HttpListenerRequest request = context.Request;
        HttpListenerResponse response = context.Response;

        // Enable CORS
        response.Headers.Add("Access-Control-Allow-Origin", "*");
        response.Headers.Add("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
        response.Headers.Add("Access-Control-Allow-Headers", "Content-Type");

        if (request.HttpMethod == "OPTIONS")
        {
            response.StatusCode = 204;
            response.Close();
            return;
        }

        try
        {
            string path = request.Url?.AbsolutePath ?? "/";
            Console.WriteLine($"{request.HttpMethod} {path}");

            if (path == "/health")
            {
                await HandleHealth(response);
            }
            else if (path == "/generate" && request.HttpMethod == "POST")
            {
                await HandleGenerate(request, response);
            }
            else if (path == "/inference.html" || path == "/")
            {
                await HandleStaticFile(response, "inference.html");
            }
            else
            {
                response.StatusCode = 404;
                byte[] buffer = Encoding.UTF8.GetBytes("Not Found");
                await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error handling request: {ex.Message}");
            response.StatusCode = 500;
            byte[] buffer = Encoding.UTF8.GetBytes($"{{\"error\": \"{ex.Message}\"}}");
            await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
        }
        finally
        {
            response.Close();
        }
    }

    private static async Task HandleHealth(HttpListenerResponse response)
    {
        response.ContentType = "application/json";
        string json = "{\"status\": \"ok\", \"model\": \"loaded\"}";
        byte[] buffer = Encoding.UTF8.GetBytes(json);
        await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
    }

    private static async Task HandleGenerate(HttpListenerRequest request, HttpListenerResponse response)
    {
        // Parse request body
        using StreamReader reader = new StreamReader(request.InputStream, request.ContentEncoding);
        string body = await reader.ReadToEndAsync();
        
        var requestData = JsonSerializer.Deserialize<GenerateRequest>(body);
        if (requestData == null || string.IsNullOrEmpty(requestData.Prompt))
        {
            response.StatusCode = 400;
            byte[] buffer = Encoding.UTF8.GetBytes("{\"error\": \"Missing prompt\"}");
            await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
            return;
        }

        // Set up SSE
        response.ContentType = "text/event-stream";
        response.Headers.Add("Cache-Control", "no-cache");
        response.Headers.Add("Connection", "keep-alive");

        Console.WriteLine($"Generating from prompt: \"{requestData.Prompt}\"");
        Console.Write("Tokens: ");

        try
        {
            int tokenCount = 0;
            foreach (var token in Transformer.GenerateStream(
                requestData.Prompt, 
                requestData.MaxTokens, 
                requestData.Temperature))
            {
                // Send SSE event
                string eventData = $"data: {JsonSerializer.Serialize(new { token })}\n\n";
                byte[] buffer = Encoding.UTF8.GetBytes(eventData);
                await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
                await response.OutputStream.FlushAsync();
                
                Console.Write(token);
                tokenCount++;
            }

            // Send done event
            string doneData = "data: {\"done\": true}\n\n";
            byte[] doneBuffer = Encoding.UTF8.GetBytes(doneData);
            await response.OutputStream.WriteAsync(doneBuffer, 0, doneBuffer.Length);
            await response.OutputStream.FlushAsync();

            Console.WriteLine($"\nGenerated {tokenCount} tokens\n");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"\nError during generation: {ex.Message}\n");
            string errorData = $"data: {{\"error\": \"{ex.Message}\"}}\n\n";
            byte[] errorBuffer = Encoding.UTF8.GetBytes(errorData);
            await response.OutputStream.WriteAsync(errorBuffer, 0, errorBuffer.Length);
        }
    }

    private static async Task HandleStaticFile(HttpListenerResponse response, string filename)
    {
        // Try to find inference.html in current directory or parent directories
        string[] searchPaths = new[]
        {
            filename,
            Path.Combine("../../cabi", filename),
            Path.Combine("../cabi", filename),
            Path.Combine("../../python/examples", filename)
        };

        string? filePath = null;
        foreach (var path in searchPaths)
        {
            if (File.Exists(path))
            {
                filePath = path;
                break;
            }
        }

        if (filePath == null)
        {
            response.StatusCode = 404;
            byte[] buffer = Encoding.UTF8.GetBytes($"File not found: {filename}");
            await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
            return;
        }

        response.ContentType = "text/html";
        byte[] content = File.ReadAllBytes(filePath);
        await response.OutputStream.WriteAsync(content, 0, content.Length);
    }
}

public class GenerateRequest
{
    public string Prompt { get; set; } = "";
    public int MaxTokens { get; set; } = 50;
    public float Temperature { get; set; } = 0.7f;
}
